elastic/kibana
# [Uptime] Enable local mTLS and deprecate basic auth for the synthetics service

> ⚠️ A rebase accidentally triggered lots of reviews, please ignore. A new PR with the correct reviewers is at: https://github.com/elastic/kibana/pull/127116

## Summary

This PR makes it much easier to authenticate to a local instance of the service using `mTLS`.

> ⚠️ Please notice that this PR still does not deprecate Basic Auth as we'll need to introduce a way of obtaining certificates for testing local instances of Kibana against remote service instances.

> 👉  I won't link the service-side PR as that's a private repository still.

This PR also includes updates which:
* Allow HTTPS requests to be sent to a server whose certificates are self-signed (when not in `production`)
* Allow users to set a `devUrl` without a `manifest` (when not in `production`), further simplifying the dev setup and thus improving DX.
* Allow users to have both a `devUrl` _and_ locations from a `manifest` when not in `production`.
* Automatically fetch `locations` on startup rather than waiting for the first scheduled task to occur
* Fix the return values from the server and the criteria to display errors. Before, we were always returning an empty array if `errors` existed, even if it were an empty array (which it always us if everything went fine).

## How to test this PR

> ⚠️ Make sure that you have the latest version (`0.9.0` or later) of the Synthetics integration installed and that indexes/templates have been updated according to that package's spec. You may need to delete existing data-sources and reinstall if you already have data.

> 👉  Testing this PR requires you to run your own local version of the service using the branch for the PR which corresponds to this one.

1. Start an ES cluster and bind it to your machine's `9200` port. You can do this using `elastic-package stack up`.
2. Checkout the corresponding branch for the `synthetics-service`, and re-generate certificates by deleting the folder with the existing ones and re-running `generate-certs.sh`.
3. Start the `synthetics-service` using the generated certificates.
    ```
    DEV=true LOGLEVEL=trace CACERTFILE=./k8s/cert-test/build/bundle.pem SERVERCERT=./k8s/cert-test/build/local-server.pem SERVERKEYFILE=./k8s/cert-test/build/local-server.key PORT=10001 ./synthetics-service
    ```
    > ☝️  Ensure that you have a local Kubernetes cluster (using Minikube) for the service to connect to.
    > You can create one by running the service's `make start` task described in the service's `k8s` folder's `Makefile`.
4. Copy the client certificates to the `config` folder, and update Kibana's configuration file:
    ```yaml
    # Enables the monitor management UI itself
    xpack.uptime.ui.monitorManagement.enabled: true
    
    # This allows the docker containers running the jobs to connect to
    # the ES instance bound to your machine's `9200` port.
    xpack.uptime.service.hosts: [http://host.docker.internal:9200]
    
    # Make sure to change the port depending on which PORT the service is running
    xpack.uptime.service.devUrl: https://localhost:10001
    
    # These are the files in `k8s/cert-test/build`.
    # In this example, they've been copied to the kibana/config folder.
    xpack.uptime.service.tls.certificate: config/123-client.pem
    xpack.uptime.service.tls.key: config/123-client.key
    ```
5. Start your Kibana instance and try to use the service as you were using before. You should be able to create, edit, and run monitors just like before.

> ⭐  Please also test setting `devUrl` and the `NODE_ENV` as `production` to ensure that `devUrl` is ignored in that case.

## Testing a local instance with a remote cluster (⚠️ Please try this)

1. Start a local Kibana instance with `basicAuth` set into its config file
2. Use NGrok to expose your Elasticsearch port `9200` (you will want to change the default username and password) and use that public URL for `xpack.uptime.service.hosts`. When adding the NGrok URL there, please specify port `80` after the NGrok URL (i.e. `xpack.uptime.service.hosts: [http://xxx-xxx-xxx.ngrok.io:80]`).
3. Set the `manifestUrl` for the desired remote service's manifest
4. Create a monitor and run it on the remote service.

You should see results both for `run_once`, and after the monitor is saved.

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials (added to the `synthetics-service` repo`)
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker) (I haven't done any changes here, but also haven't seen the previous key in this file, can someone provide input on this?)

### Risks

I thought it would be better to list risks rather than create a table as I'd like other folks to double-check them during the tests:

* Whether you can still access the service just fine without any changes to your `/etc/hosts` file
* Do you think there's a problem with the way I have disallowed `devUrl` to be used in `production`? Are there further checks we should do?
* I have updated the server to create a location for `devUrl` out of `production` when it's set, do you think that's an adequate approach?


### For maintainers


> 👉   Please notice that I didn't consider this change to be a breaking change because I'm touching pieces of software not yet GA (not even close to that actually). Therefore, I've only labelled this as a deprecation.
